drivers/serial/uart_nrfx_uarte2.c

```diff
@@ -0,0 +1,1058 @@
+	if (hwfc && timeout != SYS_FOREVER_US) {
+		k_timer_start(&data->async->tx_timer, K_USEC(timeout), K_NO_WAIT);
+	}
+
+	return 0;
+}
+
+static int api_tx_abort(const struct device *dev)
+{
+	const nrfx_uarte_t *nrfx_dev = get_nrfx_dev(dev);
+	nrfx_err_t err;
+
+	err = nrfx_uarte_tx_abort(nrfx_dev, false);
+	return (err == NRFX_SUCCESS) ? 0 : -EFAULT;
+}
+
+static void tx_timeout_handler(struct k_timer *timer)
+{
+	const struct device *dev = (const struct device *)k_timer_user_data_get(timer);
+
+	(void)api_tx_abort(dev);
+}
+
+static void rx_timeout_handler(struct k_timer *timer)
+{
+	const struct device *dev = (const struct device *)k_timer_user_data_get(timer);
+	struct uarte_nrfx_data *data = dev->data;
+	struct uarte_async_data *adata = data->async;
+	const nrfx_uarte_t *nrfx_dev = get_nrfx_dev(dev);
+
+	if (nrfx_uarte_rx_new_data_check(nrfx_dev)) {
+		adata->idle_cnt = RX_TIMEOUT_DIV - 1;
+	} else {
+		adata->idle_cnt--;
+		if (adata->idle_cnt == 0) {
+			(void)nrfx_uarte_rx_abort(nrfx_dev, false, false);
+			return;
+		}
+	}
+
+	start_rx_timer(data);
+}
+
+/* Determine if RX FIFO content shall be kept when device is being disabled.
+ * When flow-control is used then we expect to keep RX FIFO content since HWFC
+ * enforces lossless communication. However, when HWFC is not used (by any instance
+ * then RX FIFO handling feature is disabled in the nrfx_uarte to save space.
+ * It is based on assumption that without HWFC it is expected that some data may
+ * be lost and there are means to prevent that (keeping receiver always opened by
+ * provided reception buffers on time).
+ */
+static inline uint32_t get_keep_fifo_content_flag(const struct device *dev)
+{
+	struct uarte_nrfx_data *data = dev->data;
+	const struct uarte_nrfx_config *config = dev->config;
+
+	if ((IS_ENABLED(CONFIG_UART_USE_RUNTIME_CONFIGURE) &&
+	     (data->uart_config.flow_ctrl == UART_CFG_FLOW_CTRL_RTS_CTS)) ||
+	     (config->config.flow_ctrl == UART_CFG_FLOW_CTRL_RTS_CTS)) {
+		return NRFX_UARTE_RX_ENABLE_KEEP_FIFO_CONTENT;
+	}
+
+	return 0;
+}
+
+static int api_rx_enable(const struct device *dev, uint8_t *buf, size_t len, int32_t timeout)
+{
+	nrfx_err_t err;
+	const nrfx_uarte_t *nrfx_dev = get_nrfx_dev(dev);
+	struct uarte_nrfx_data *data = dev->data;
+	struct uarte_async_data *adata = data->async;
+	uint32_t flags = NRFX_UARTE_RX_ENABLE_CONT |
+			 get_keep_fifo_content_flag(dev) |
+			 (is_async_api(dev) ? NRFX_UARTE_RX_ENABLE_STOP_ON_END : 0);
+
+	if (timeout != SYS_FOREVER_US) {
+		/* Set minimum interval to 3 RTC ticks. 3 is used due to RTC limitation
+		 * which cannot set timeout for next tick. Assuming delay in processing
+		 * 3 instead of 2 is used. Note that lower value would work in a similar
+		 * way but timeouts would always occur later than expected,  most likely
+		 * after ~3 ticks.
+		 */
+		adata->idle_cnt = RX_TIMEOUT_DIV + 1;
+		adata->rx_timeout = K_USEC(timeout / RX_TIMEOUT_DIV);
+		adata->rx_timeout.ticks = MAX(3, adata->rx_timeout.ticks);
```

Comment:
We got rid of such rounding up to 3 ticks in 52c55177ba71af328c6069ea56c007e6eb5cdf1d. Why is it coming back? See also https://github.com/nrfconnect/sdk-zephyr/pull/1336#discussion_r1360443426. Kernel ticks may not be equal to RTC ticks, and for GRTC the correction is completely inappropriate.

Reply:
Removed.